File: Code-Eval/moderate/valid_parenthesis.py
# Python 3
import sys

match = {'(': ')', '{': '}', '[': ']'}
test_cases = open(sys.argv[1], 'r')
for test in test_cases:
    if test == '':
        continue
    parens = list(test.strip())
    pStack = []
    answer = "True"
    
    for p in parens:
        if p == '(' or p == '{' or p == '[':
            pStack.append(p)
        elif p == ')' or p == '}' or p == ']':
            if len(pStack) == 0 or p != match[pStack[-1]]:
                #print(p, pStack)
                answer = "False"
                break
            else:
                pStack.pop()
    if len(pStack) != 0:
        answer = "False"
    print(answer)
    

test_cases.close()

File: Code-Eval/moderate/valid_parenthesis.lua
match = {["("] = ")", ["{"] = "}", ["["] = "]"}

for parens in io.lines(arg[1]) do
  pStack = {}
  answer = "True"

  for i = 1, # parens do
    local p = parens:sub(i,i)
    --print(parens, type(parens), p)
    if p == "(" or p == "{" or p == "[" then
      table.insert(pStack, p)
    elseif p == ")" or p == "}" or p == "]" then
      if # pStack == 0 or p ~= match[pStack[# pStack]] then
        answer = "False"
        break
      else
        table.remove(pStack)
      end
    end
  end

  if # pStack ~= 0 then
    answer = "False"
  end
  print(answer)
end
